# Partie 0 installation
Pour traiter des données, nous allons utiliser la bibliothèque Python Pandas. Une bibliothèque Python permet de rajouter des fonctionnalités par rapport au langage de base. La bibliothèque Pandas est donc très utilisée pour tout ce qui touche au traitement des données.
 Pour installer pandas, faire 


In [36]:
!  pip install pandas


Pour vérifier que pandas fonctionne bien, veuillez lancer la cellulle suivante


In [37]:
import pandas
def lecture():
    iden=pandas.read_csv("test_pandas.csv")
    return iden
lecture()

Message1           Message2
0  Pandas est ok   tout est parfait
1  Pandas est ok   tout est parfait
2  Pandas est ok   tout est parfait

# Partie 1 Utiliser le module pandas

Commençons ! 

Pour nos premiers pas avec Pandas, nous allons utiliser des données très simples au format CSV : 
ces données sont contenues dans le fichier ident_virgule.csv que vous trouverez à la racine de votre dossier.

Le code ci-dessous est très simple :

- Avec la première ligne, nous importons la bibliothèque pandas afin de pouvoir l'utiliser
- À la deuxième ligne,nous créons une nouvelle fonction nommée lecture(). La création s'effeectue avec le mot clé `def`
- A la troisième ligne, dans le sous programme de la fonction, nous créons une variable "iden" qui va contenir les données présentes dans le fichier ident_virgule.csv"
- A la 4ème ligne,toujours dans le sous programme de la fonction, on renvoie la variable iden

Fin de la focntion et retour à la ligne
- à la dernière ligne, la fonction lecture est lancée.

A la troisième ligne, nous demandons à python d'afficher le contenu de la variable iden.

Exécutez la cellule ci dessous pour tester ce code.


In [38]:
import pandas
def lecture():
    iden=pandas.read_csv("ident_virgule.csv")
    return(iden)

lecture()

nom       prenom date_naissance
0  Durand  Jean-Pierre     23/05/1985
1  Dupont   Christophe     15/12/1967
2   Terta        Henry     12/06/1978

 Vous avez peut-être remarqué qu'une colonne a été ajoutée par rapport à ce que nous obtenions avec le tableur :

Les nombres présents dans cette colonne sont appelés des index.
Chaque ligne du tableau a un index (première ligne : index 0, deuxième ligne index 1...)

ATTENTION : les index commencent à 0 et pas à 1

Les colonnes possèdes également des index, dans notre exemple ces index correspondent au "nom" (index de la première colonne), au "prenom" (index de la deuxième colonne) et à "date_naissance" (index de la troisième colonne)

En résumé : 
- les lignes possèdent des index (0,1,2..).
- les colonnes possèdent aussi des index ("nom", "prenom",...)

Il est possible de récupérer certaines données du tableau, par exemple, certaines lignes, certaines colonnes ou bien encore des valeurs uniques. 

Pour cela, il suffit d'utiliser l'instruction `loc` avec les index des lignes et les index des colonnes. 

Le principe de fonctionnement de "loc" est relativement simple puisque l'on aura une instruction de la forme 
loc[index_ligne,index_colonne]



# À faire vous-même 5
Testez le programme suivant  et étudier attentivement la ligne 4

In [39]:
import pandas
def lecture():
    iden=pandas.read_csv("ident_virgule.csv")
    info=iden.loc[1,'prenom']
    return info

lecture()

'Christophe'

# À faire vous-même 6

Modifiez le programme du "À faire vous-même 5" pour que la variable info renvoyée contienne "12/06/1978" 

In [40]:
import pandas
def lecture():
    iden=pandas.read_csv("ident_virgule.csv")
    info=iden.loc[1,'prenom']
    return info

lecture()

'Christophe'

 Il est possible de récupérer toutes les lignes d'une colonne, il suffit de remplacer la partie " index_ligne " de `loc` par **" : "**

# À faire vous-même 7

Testez le programme suivant :

Vérifiez que la variable "info" contient bien toutes les données de la colonne d'index "nom", autrement dit, tous les noms 

In [41]:
import pandas
def lecture():
    iden=pandas.read_csv("ident_virgule.csv")
    info=iden.loc[:,'nom']
    return info

lecture()

0    Durand
1    Dupont
2     Terta
Name: nom, dtype: object

Il est possible de récupérer toutes les colonnes d'une ligne particulière, cette fois en remplaçant la partie "index_colonne" de "loc" par ":"

# À faire vous-même 8

Testez le programme suivant : 

Vérifiez que la variable "info" contient bien toutes les données de la dernière ligne (index 2) 

In [42]:
import pandas
def lecture():
    iden=pandas.read_csv("ident_virgule.csv")
    info=iden.loc[2,:]
    return info

lecture()


nom                    Terta
prenom                 Henry
date_naissance    12/06/1978
Name: 2, dtype: object

Il est aussi possible de récupérer seulement certaines lignes et certaines colonnes en utilisant la notation suivante :

loc[[index_ligne_1,index_ligne_2,...],[index_colonne_1,index_colonne_2,...]] 

# À faire vous-même 9

Testez le programme suivant : 

Vérifiez que la variable "info" contient bien un tableau avec uniquement 
- les colonnes "nom" et "date_naissance" 
- issuees de la première ligne (index 0) et de la deuxième ligne (index 1). 

In [43]:

import pandas
def lecture():
    iden=pandas.read_csv("ident_virgule.csv")
    info=iden.loc[[0,1],['nom','date_naissance']]
    return info

lecture()



nom date_naissance
0  Durand     23/05/1985
1  Dupont     15/12/1967

# Partie 2
Afin d'avoir des exemples plus complexes à traiter, dans la suite, nous allons travailler sur les données contenues dans le fichier ville_virgule.csv. 

# À faire vous-même 10

Testez le programme suivant : 


In [44]:
import pandas
def lecture():
    info=pandas.read_csv("villes_virgule.csv")
    return info

lecture()


dep                       nom     cp  nb_hab_2010  nb_hab_1999  \
0        1                      Ozan   1190          618          469   
1        1     Cormoranche-sur-Saône   1290         1058          903   
2        1                    Plagne   1130          129           83   
3        1                   Tossiat   1250         1406         1111   
4        1                  Pouillat   1250           88           58   
...    ...                       ...    ...          ...          ...   
36695  976                      Sada  97640        10195        10195   
36696  976                  Tsingoni  97680        10454        10454   
36697  971          Saint-Barthélemy  97133         8938         8938   
36698  971              Saint-Martin  97150        36979        36979   
36699  975  Saint-Pierre-et-Miquelon  97500         6080         6080   

       nb_hab_2012  dens    surf      long       lat  alt_min  alt_max  
0              500    93    6.60   4.91667  46.38330    170.0    205.0  
1             1000   107    9.85   4.83333  46.23330    168.0    211.0  
2              100    20    6.20   5.73333  46.18330    560.0    922.0  
3             1400   138   10.17   5.31667  46.13330    244.0    501.0  
4              100    14    6.23   5.43333  46.33330    333.0    770.0  
...            ...   ...     ...       ...       ...      ...      ...  
36695        10195   933   10.92  45.10470 -12.84860      NaN      NaN  
36696        10454   300   34.76  45.10700 -12.78970      NaN      NaN  
36697         8938   372   24.00 -62.83330  17.91670      NaN      NaN  
36698        36979   695   53.20  18.09130 -63.08290      NaN      NaN  
36699         6080    25  242.00  46.71070   1.71819      NaN      NaN  

[36700 rows x 12 columns]

Comme vous pouvez le constater, il manque des données dans le tableau qui s'affiche dans la console (les données manquantes sont symbolisées par des ...), en effet, le tableau contient trop données pour qu'il soit entièrement affiché dans la console .

En effet ce fichier contient 36700 lignes


# À faire vous-même 11

En explorant le tableau, vous devriez, notamment dans les colonnes l'altitude mini et maxi, voir apparaitre un étrange "nan" pour les dernières villes du tableau. "nan" signifie "not a number", ici, cela veut tout simplement dire que certaines données sont manquantes.

Nous allons maintenant introduire des conditions dans la sélection des villes. 
Imaginez par exemple que vous désirez obtenir un tableau contenant toutes les villes qui ont une altitude minimum supérieure à 1500 m :


Analysez et testez le programme suivant : 




In [45]:
import pandas
def lecture():
    info_villes=pandas.read_csv("villes_virgule.csv")
    nom_alt=info_villes.loc[info_villes["alt_min"]>1500,["nom","alt_min"]]
    return nom_alt

lecture()


nom  alt_min
1618                Larche   1606.0
1790              Ristolas   1571.0
1798           Saint-Véran   1756.0
1847    Molines-en-Queyras   1625.0
1904                Abriès   1513.0
1923        Villar-d'Arêne   1519.0
26927          La Llagonne   1546.0
26943  Caudiès-de-Conflent   1616.0
27039      Porté-Puymorens   1557.0
27125           Mont-Louis   1516.0
27134               Angles   1531.0
29970              Bessans   1673.0
30114          Val-d'Isère   1785.0
30127     Bonneval-sur-Arc   1759.0

Dans le "loc", l'expression "info_villes["alt_min"]>1500" est bien avant la virgule, elle concerne donc les index des lignes du tableau. On sélectionnera uniquement les lignes qui auront la valeur du descripteur "alt_min" supérieure à 1500. Nous allons donc bien sélectionner les villes qui ont une altitude minimum supérieure à 1500 m

# À faire vous-même 12

En vous inspirant de ce qui a été fait au "À faire vous-même 11", écrivez un programme qui permettra d'avoir les villes qui ont une densité d'habitant inférieure strictement à 1 (dans le tableau ainsi créé, on aura 3 colonnes : le nom de la ville, la densité de la population et l'altitude minimum) 

Modifier donc le programme ci dessous pour connaitre ces villes fantômes de France:

In [46]:
import pandas
def lecture():
    info_villes=pandas.read_csv("villes_virgule.csv")
    nom_alt=info_villes.loc[info_villes["alt_min"]>1500,["nom","alt_min"]]
    return nom_alt

lecture()

nom  alt_min
1618                Larche   1606.0
1790              Ristolas   1571.0
1798           Saint-Véran   1756.0
1847    Molines-en-Queyras   1625.0
1904                Abriès   1513.0
1923        Villar-d'Arêne   1519.0
26927          La Llagonne   1546.0
26943  Caudiès-de-Conflent   1616.0
27039      Porté-Puymorens   1557.0
27125           Mont-Louis   1516.0
27134               Angles   1531.0
29970              Bessans   1673.0
30114          Val-d'Isère   1785.0
30127     Bonneval-sur-Arc   1759.0

Il est possible de combiner plusieurs facteurs de sélection en utilisant 
- un "et"  que l'on code avec une esperluette `&` 
ou 
- un "ou"  que l'on code avec un tube  `|`    (<kbd>AltGr</kbd> +  <kbd>6</kbd>)

# À faire vous-même 13

Analysez et testez le programme suivant : 

In [47]:
import pandas

def lecture():
    info_villes=pandas.read_csv("villes_virgule.csv")
    nom_alt=info_villes.loc[(info_villes["alt_min"]>1500) & (info_villes["dens"]>50),["nom","dens","alt_min"]]
    return nom_alt

lecture()



nom  dens  alt_min
27125  Mont-Louis   633   1516.0

Vous devriez constater qu'il y a, en France, une seule ville avec une densité de population supérieure à 50 et une altitude minimum supérieure à 1500 m.

Il est aussi possible d'effectuer des calculs sur des colonnes, par exemple des moyennes. Il suffit d'utiliser l'instruction "mean" pour effectuer une moyenne :

# À faire vous-même 14

Analysez et testez le programme suivant : 

In [48]:
import pandas

def lecture():
    info_villes=pandas.read_csv("villes_virgule.csv")
    moyenne_alt_min=info_villes.loc[:,"alt_min"].mean()
    return moyenne_alt_min

lecture()


193.15756945963685

Vous devriez constater que l'altitude minimum moyenne est de 193 m en France. 

Je rappelle que dans "loc[:,"alt_min"]", le ":" signifie que l'on considère toutes les lignes du tableau. 

De plus le "alt_min" que le calcul de la moyenne porte bien sur les données du descripteur "alt_min".

# À faire vous-même 15

Modifier le programme du "À faire vous-même 14" qui a été recopié dans la cellule suivante.
Vous dever calculer le nombre moyen d'habitants en 2012 :




In [49]:
import pandas

def lecture():
    info_villes=pandas.read_csv("villes_virgule.csv")
    moyenne_alt_min=info_villes.loc[:,"alt_min"].mean()
    return moyenne_alt_min

lecture()


193.15756945963685

Pour l'instant nous avons calculé une moyenne sur l'ensemble des lignes, il est aussi possible d'imposer une condition sur les lignes qui seront utilisées pour le calcul.

# À faire vous-même 16

Analysez et testez le programme suivant : 

In [50]:
import pandas

def lecture():
    info_villes=pandas.read_csv("villes_virgule.csv")
    nbe_hab=info_villes.loc[info_villes["alt_min"]>1500,"nb_hab_2012"].mean()
    return nbe_hab

lecture()



350.0

Vous devriez constater que les villes ayant une altitude minimum supérieure à 1500 m avaient en moyenne 350 habitants en 2012.
Il est aussi possible de trier le tableau en fonction des valeurs d'un descripteur. Il suffit d'utiliser l'instruction `sort_values`

# À faire vous-même 17

Analysez et testez le programme suivant : 

In [51]:
import pandas

def lecture():
    info_villes=pandas.read_csv("villes_virgule.csv")
    tri_alt_min=info_villes.sort_values(by=["alt_min"])
    return tri_alt_min

lecture()


dep                       nom     cp  nb_hab_2010  nb_hab_1999  \
11100   29                   Quimper  29000        63550        63274   
21570   56                   Le Hézo  56450          730          551   
21409   56            Île-aux-Moines  56780          629          613   
21466   56                     Baden  56870         4137         3370   
10963   29                Treffiagat  29730         2378         2171   
...    ...                       ...    ...          ...          ...   
36695  976                      Sada  97640        10195        10195   
36696  976                  Tsingoni  97680        10454        10454   
36697  971          Saint-Barthélemy  97133         8938         8938   
36698  971              Saint-Martin  97150        36979        36979   
36699  975  Saint-Pierre-et-Miquelon  97500         6080         6080   

       nb_hab_2012  dens    surf      long       lat  alt_min  alt_max  
11100        63900   752   84.45  -4.10000  48.00000     -5.0    151.0  
21570          700   149    4.89  -2.70000  47.58330     -1.0     38.0  
21409          600   196    3.20   1.15528  47.59670     -1.0     31.0  
21466         4000   175   23.53  -2.91667  47.61670     -1.0     43.0  
10963         2300   293    8.10  -4.26667  47.81670     -1.0     26.0  
...            ...   ...     ...       ...       ...      ...      ...  
36695        10195   933   10.92  45.10470 -12.84860      NaN      NaN  
36696        10454   300   34.76  45.10700 -12.78970      NaN      NaN  
36697         8938   372   24.00 -62.83330  17.91670      NaN      NaN  
36698        36979   695   53.20  18.09130 -63.08290      NaN      NaN  
36699         6080    25  242.00  46.71070   1.71819      NaN      NaN  

[36700 rows x 12 columns]

Vous devriez obtenir un nouveau tableau de données "tri_alt_min" trié dans l'ordre croissant des altitudes minimums. Quelle est la ville ayant l'altitude minimum la plus faible de France ?

Il est aussi possible de trier par ordre décroissant en ajoutant `ascending=False` :

# À faire vous-même 18

Analysez et testez le programme suivant : 

In [52]:
import pandas

def lecture():
    info_villes=pandas.read_csv("villes_virgule.csv")
    tri_alt_min=info_villes.sort_values(by=["alt_min"], ascending=False)
    return tri_alt_min

lecture()


dep                       nom     cp  nb_hab_2010  nb_hab_1999  \
30114   73               Val-d'Isère  73150         1563         1628   
30127   73          Bonneval-sur-Arc  73480          241          239   
1798     5               Saint-Véran   5350          257          265   
29970   73                   Bessans  73480          343          310   
1847     5        Molines-en-Queyras   5350          315          322   
...    ...                       ...    ...          ...          ...   
36695  976                      Sada  97640        10195        10195   
36696  976                  Tsingoni  97680        10454        10454   
36697  971          Saint-Barthélemy  97133         8938         8938   
36698  971              Saint-Martin  97150        36979        36979   
36699  975  Saint-Pierre-et-Miquelon  97500         6080         6080   

       nb_hab_2012  dens    surf      long       lat  alt_min  alt_max  
30114         1600    16   94.39   6.98333  45.45000   1785.0   3599.0  
30127          200     2   82.72   7.05000  45.36670   1759.0   3642.0  
1798           300     5   44.75   6.86667  44.70000   1756.0   3175.0  
29970          300     2  128.08   6.99167  45.31670   1673.0   3754.0  
1847           300     5   53.62   6.85000  44.73330   1625.0   3160.0  
...            ...   ...     ...       ...       ...      ...      ...  
36695        10195   933   10.92  45.10470 -12.84860      NaN      NaN  
36696        10454   300   34.76  45.10700 -12.78970      NaN      NaN  
36697         8938   372   24.00 -62.83330  17.91670      NaN      NaN  
36698        36979   695   53.20  18.09130 -63.08290      NaN      NaN  
36699         6080    25  242.00  46.71070   1.71819      NaN      NaN  

[36700 rows x 12 columns]

Quelle est la ville ayant l'altitude minimum la plus importante de France ?

# À faire vous-même 19

Modifier le programme ci dessous afin de répondre à la question suivante : 
quelle est la ville ayant la densité de population la plus forte ? 



In [53]:
import pandas

def lecture():
    info_villes=pandas.read_csv("villes_virgule.csv")
    tri_alt_min=info_villes.sort_values(by=["alt_min"], ascending=False)
    return tri_alt_min

lecture()

dep                       nom     cp  nb_hab_2010  nb_hab_1999  \
30114   73               Val-d'Isère  73150         1563         1628   
30127   73          Bonneval-sur-Arc  73480          241          239   
1798     5               Saint-Véran   5350          257          265   
29970   73                   Bessans  73480          343          310   
1847     5        Molines-en-Queyras   5350          315          322   
...    ...                       ...    ...          ...          ...   
36695  976                      Sada  97640        10195        10195   
36696  976                  Tsingoni  97680        10454        10454   
36697  971          Saint-Barthélemy  97133         8938         8938   
36698  971              Saint-Martin  97150        36979        36979   
36699  975  Saint-Pierre-et-Miquelon  97500         6080         6080   

       nb_hab_2012  dens    surf      long       lat  alt_min  alt_max  
30114         1600    16   94.39   6.98333  45.45000   1785.0   3599.0  
30127          200     2   82.72   7.05000  45.36670   1759.0   3642.0  
1798           300     5   44.75   6.86667  44.70000   1756.0   3175.0  
29970          300     2  128.08   6.99167  45.31670   1673.0   3754.0  
1847           300     5   53.62   6.85000  44.73330   1625.0   3160.0  
...            ...   ...     ...       ...       ...      ...      ...  
36695        10195   933   10.92  45.10470 -12.84860      NaN      NaN  
36696        10454   300   34.76  45.10700 -12.78970      NaN      NaN  
36697         8938   372   24.00 -62.83330  17.91670      NaN      NaN  
36698        36979   695   53.20  18.09130 -63.08290      NaN      NaN  
36699         6080    25  242.00  46.71070   1.71819      NaN      NaN  

[36700 rows x 12 columns]